### ISSUE TITLE
TcpStream::connect_timeout doesn't fail to connect to a nonexistent server on Linux

### ISSUE THREAD

**Issue by sfackler** [rust-lang/rust#45265]
```rust
use std::net::TcpStream;
use std::time::Duration;

fn main() {
    let sock = "127.0.0.1:1234".parse().unwrap();
    let timeout = Duration::from_secs(3);
    TcpStream::connect_timeout(&sock, timeout).unwrap();
}
```

This program succeeds on Linux, even though it shouldn't (assuming you don't have a listener on 1234). Trying to do anything with the socket produces `Error { repr: Os { code: 107, message: "Transport endpoint is not connected" } }`.

**Comment by sfackler** [comment]
Interestingly it does behave as expected on macOS.

**Comment by karizon** [comment]
Works under WSL as well.  Native linux it fails spectacularly.

**Comment by sfackler** [comment]
From an strace, it looks like Linux sets all of `POLLOUT`, `POLLERR`, and `POLLHUP` at once. We treat `POLLOUT` as success, but it seems like we should be checking for `POLLERR` instead:

```
socket(AF_INET, SOCK_STREAM|SOCK_CLOEXEC, IPPROTO_IP) = 3
ioctl(3, FIONBIO, [1])                  = 0
connect(3, {sa_family=AF_INET, sin_port=htons(80), sin_addr=inet_addr("127.0.0.1")}, 16) = -1 EINPROGRESS (Operation now in progress)
ioctl(3, FIONBIO, [0])                  = 0
clock_gettime(CLOCK_MONOTONIC, {tv_sec=632431, tv_nsec=149012221}) = 0
clock_gettime(CLOCK_MONOTONIC, {tv_sec=632431, tv_nsec=149240049}) = 0
poll([{fd=3, events=POLLOUT}], 1, 2999) = 1 ([{fd=3, revents=POLLOUT|POLLERR|POLLHUP}])
```

**Comment by sfackler** [comment]
Fix here: #45269
